packages/rich-text/src/editor.tsx

```diff
@@ -18,22 +18,30 @@
  * External dependencies
  */
 import { Editor, getDefaultKeyBinding, KeyBindingUtil } from 'draft-js';
-import PropTypes from 'prop-types';
 import {
   useEffect,
   useRef,
   useImperativeHandle,
   forwardRef,
   useUnmount,
 } from '@googleforcreators/react';
+import type { ForwardedRef, KeyboardEvent } from 'react';
 
 /**
  * Internal dependencies
  */
 import useRichText from './useRichText';
 import customInlineDisplay from './customInlineDisplay';
 
-function RichTextEditor({ content, onChange }, ref) {
+interface RichTextEditorProps {
+  content: string;
+  onChange: (content: string) => void;
+}
+
+function RichTextEditor(
+  { content, onChange }: RichTextEditorProps,
+  ref: ForwardedRef<HTMLElement>
+) {
   const editorRef = useRef(null);
   const {
     state: { editorState },
```

Comment:
@barklund Do you have any thoughts on what's the correct typing approach here with the `useRichText`, state and the `useContextSelector`: https://github.com/GoogleForCreators/web-stories-wp/blob/910c86e4114ced8b5e80b5abfec76627e6874e54/packages/react/src/useContextSelector.ts#L47-L51

Since `useContextSelector` return `T | undefined`, it complains about `state` and `action` not existing in the return value of `useRichText`.
Additionally, using just `useRichText()` is complaining about not having enough arguments, however, making the `selector` optional in `useRichText` in will cause new type errors with using `useContextSelector` inside it, so I'm wondering if `useContextSelector` is still not with correct types.

Let me know if you have any insights here 🤓

Reply:
If we change `useContextSelector` line 56 to:

```ts
    if (ref.current && equalityFn(ref.current, selected)) {
```

This will change the return to `S` rather than `S | undefined`. I think that's a reasonable change.

If it's `falsey` (most likely some simple value), we don't actually care about equality anymore. This won't change any functionality as far as I can see, but would make typings easier.

@swissspidy, WDYT?

Reply:
When returning just `S`, it still doesn't allow using the context without props, e.g. just `useRichText()` gives an error of expecting at least 1 argument, making the selector optional, however, generates new type errors. So I'm wondering if `identity` also still needs adjustments, too, to allow using it without `selector`.
https://github.com/GoogleForCreators/web-stories-wp/blob/80a5cc6f593e7d8f37f656dd508759774dc3d5c6/packages/react/src/context.ts#L19

Reply:
Uh, neat! I just read about the function overloads a few days ago but didn't make the connection with this issue at all. Thanks! 🧙

Reply:
@barklund I've added the function overload, now there's no need for casting `useRichText()` 🎉 
However, using
```
function useRichText<T>(selector?: (state: State) => T) {
  return useContextSelector(Context, selector ?? identity<State>);
}
```
still complains:
![Screenshot 2022-09-12 at 17 44 29](https://user-images.githubusercontent.com/3294597/189684179-68b9bcf7-0959-42ce-ba8b-ae4dc11ffe1d.png)

Let me know if you changed anything else when looking into this since it worked for you.

Reply:
Hmm, yeah I didn't get this error. I'll check it out and see what I perhaps did differently?

Reply:
Hmm, I can't get it to compile now either. I swore that it worked (or at least seemed to work), but that probably says more about VSCode's flaky TS compilation than the code being correct. I don't know how to correctly type this then, but I'm sure it's possible! I'll ask around.

In the meantime, I guess we have to bite the bullet and make two different functions:

```ts
export function useFullRichText(): State {
  return useContextSelector(Context, identity);
}

export function useRichText<T>(selector: (state: State) => T) {
  return useContextSelector(Context, selector);
}
```

It's a lot less elegant, but for now I guess it'll do...?

Reply:
The whole `Partial<State> | State` return type won't work, because we don't always return partial states from our selectors. We might select a single property or even computed properties, e.g. just `useRichText(({ state: { something } }) => something);`.

Reply:
TY! Using that for now.